scrapy/pipelines/files.py

```diff
@@ -645,7 +645,7 @@ def media_downloaded(
     ) -> FileInfo:
         referer = referer_str(request)
 
-        if response.status != 200:
+        if not 200 <= response.status < 300:
```

Comment:
This doesn't seem to solve the part about getting the URL from the `Location` header.